src/devices/machine/bl_handhelds_menucontrol.cpp

```diff
@@ -0,0 +1,214 @@
+		}
+	}
+	else if (m_menustate == MENU_COMMAND_00_IN)
+	{
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+	else if (m_menustate == MENU_COMMAND_01_IN)
+	{
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+	else if (m_menustate == MENU_COMMAND_02_IN)
+	{
+		m_menupos = (m_menupos & 0xff00) | ((command - 0x8) & 0xff);
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+	else if (m_menustate == MENU_COMMAND_03_IN)
+	{
+		m_menupos = (m_menupos & 0x00ff) | (((command - 0x9) & 0xff) << 8);
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+	else if (m_menustate == MENU_COMMAND_04_IN)
+	{
+		// used if you try to scroll up or left past 0 and the value becomes too large (a negative number)
+		if (m_is_unsp_type_hack)
+		{
+			if (command == 0x0d)
+				m_menupos += 4;
+			else if (command == 0x0a)
+				m_menupos += 0;
+			// used if you try to scroll up or left past 0 and the value becomes too large (a negative number)
+			// actually writes 0x314 split into 2 commands, so the 2nd write to 0x04 with param then instead 0b/16 sequence of writes instead of 26/0c adds to the high byte?
+			else if (command == 0x1e)
+				m_menupos += 0x310;
+		}
+		else
+		{
+			m_menupos += (command - 0x09);
+			m_menustate = MENU_READY_FOR_COMMAND;
+		}
+
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+	else if (m_menustate == MENU_COMMAND_05_IN)
+	{
+		// used if you try to scroll down past the and the value becomes too large
+		if (m_is_unsp_type_hack)
+		{
+			// actually writes 0x313 split into 2 commands, so the 2nd write to 0x05 with param then instead 0b/16 sequence of writes instead of 26/0c subtracts from the high byte?
+			if (command == 0x0b)
+			{
+				m_menupos -= 0xdc;
+			}
+			else if (command == 0x0e)
+			{
+				m_menupos -= 0x314;
+			}
+		}
+		else
+		{
+			if (command != 0xf0)
+				m_menupos -= (command - 0x0b);
+			else
+				m_menupos -= (command - 0x0a); // dphh8630 when pressing 'right' on final menu page
+		}
+
+		m_menustate = MENU_READY_FOR_COMMAND;
+	}
+}
+
+WRITE_LINE_MEMBER(bl_handhelds_menucontrol_device::clock_w)
+{
+	if (state)
+	{
+		m_clockstate = 1;
+	}
+	else
+	{
+		m_clockstate = 0;
+		m_command <<= 1;
+		m_command |= ((m_commandbit));
+		m_responsebit = (m_response >> (7 - m_datashifterpos)) & 1;
+		m_datashifterpos++;
+
+		if (m_datashifterpos == 8)
+		{
+			m_datashifterpos = 0;
+			handle_command();
+		}
+	}
+}
+
+WRITE_LINE_MEMBER(bl_handhelds_menucontrol_device::data_w)
+{
+	m_commandbit = state;
+}
+
+WRITE_LINE_MEMBER(bl_handhelds_menucontrol_device::reset_w)
+{
+	m_datashifterpos = 0;
+}
+
+void bl_handhelds_menucontrol_device::device_start()
+{
+	m_responsebit = 0;
+	m_clockstate = 0;
+	m_command = 0;
+	m_datashifterpos = 0;
+	m_menupos = 0;
+	m_response = 0;
+	m_commandbit = 0;
+	m_menustate = MENU_READY_FOR_COMMAND;
+
+	save_item(NAME(m_menupos));
+	save_item(NAME(m_clockstate));
+	save_item(NAME(m_datashifterpos));
+	save_item(NAME(m_responsebit));
+	save_item(NAME(m_response));
+	save_item(NAME(m_commandbit));
+	save_item(NAME(m_command));
```

Comment:
A bunch of these don’t have explicitly sized types.  You aren’t supposed to save anything that doesn’t have an explicitly sized type.

Reply:
In that case, the save system should prohibit the saving of those types, either via a static assert or a run-time fatalerror. If a driver author is able to invoke saving of these non-explicitly-sized types without either a compile failure or runtime failure, then it's something that needs to be addressed by the save system.

It shouldn't be expected that devs would have the same intimate knowledge of the save system's inner workings as you do, @cuavas.

Reply:
There isn’t a way to make it cause a compile error if the type you’re saving happens to be the type your compiler uses internally for one of the explicitly sized types.

For example on Windows/macOS i686/x86-64 or Linux i686 you’ll get no compile error for trying to save a `long long int` because the compiler uses it internally as `int64_t`.  However you’ll get a compile error for trying to save a `long long int` on Linux x86-64 because it’s not used internally for any explicitly sized integer type (`int64_t` is implemented as `long int` on Linux x86-64).  Across the systems we regularly build for, we can catch `long int`, `unsigned long int`, `long long int`, `unsigned long long int` and `char`.

Unfortunately, `int`, `unsigned int`, `short`, `signed short`, `signed char` and `unsigned char` are used to implement explicitly-sized types on all our regular target platforms.  We’d have to build for an exotic ILP64 target (e.g. Cray) to catch it at compile time.

Reply:
I was unaware of this, and it's the first time it's been mentioned